File: benchmarks/go/05_array_processing.go
package main

import (
	"fmt"
	"sort"
	"time"
)

func benchmarkMapFilterReduce(size int) {
	data := make([]int, size)
	for i := 0; i < size; i++ {
		data[i] = i
	}

	start := time.Now()

	filtered := make([]int, 0, size)
	for _, x := range data {
		if x%3 != 0 {
			filtered = append(filtered, x)
		}
	}

	mapped := make([]int, len(filtered))
	for i, x := range filtered {
		mapped[i] = x * x
	}

	result := 0
	for _, x := range mapped {
		result += x
	}

	elapsed := time.Since(start).Seconds() * 1000
	fmt.Printf("  map/filter/reduce (%d items): %.6fms, result=%d\n", size, elapsed, result)
}

func benchmarkSort(size int) {
	data := make([]int, size)
	for i := 0; i < size; i++ {
		data[i] = size - i
	}

	start := time.Now()
	sort.Ints(data)
	elapsed := time.Since(start).Seconds() * 1000

	fmt.Printf("  sort (%d items): %.6fms, first=%d, last=%d\n", size, elapsed, data[0], data[size-1])
}

func benchmarkFind(size int) {
	data := make([]int, size)
	for i := 0; i < size; i++ {
		data[i] = i
	}
	target := size - 1

	start := time.Now()
	found := 0
	for i := 0; i < 100; i++ {
		for _, x := range data {
			if x == target {
				found++
				break
			}
		}
	}
	elapsed := time.Since(start).Seconds() * 1000

	fmt.Printf("  find x100 (%d items): %.6fms, found=%d\n", size, elapsed, found)
}

func main() {
	fmt.Println("BENCHMARK: array_processing")

	benchmarkMapFilterReduce(100000)
	benchmarkMapFilterReduce(1000000)
	benchmarkSort(100000)
	benchmarkSort(1000000)
	benchmarkFind(100000)
	benchmarkFind(1000000)
}

File: benchmarks/python/05_array_processing.py
import time
from functools import reduce

def benchmark_map_filter_reduce(size):
    data = list(range(size))

    start = time.perf_counter()

    filtered = list(filter(lambda x: x % 3 != 0, data))
    mapped = list(map(lambda x: x * x, filtered))
    result = reduce(lambda acc, x: acc + x, mapped, 0)

    elapsed = (time.perf_counter() - start) * 1000
    print(f"  map/filter/reduce ({size} items): {elapsed}ms, result={result}")

def benchmark_sort(size):
    data = list(range(size, 0, -1))

    start = time.perf_counter()
    sorted_data = sorted(data)
    elapsed = (time.perf_counter() - start) * 1000

    print(f"  sort ({size} items): {elapsed}ms, first={sorted_data[0]}, last={sorted_data[-1]}")

def benchmark_find(size):
    data = list(range(size))
    target = size - 1

    start = time.perf_counter()
    found = 0
    for _ in range(100):
        r = next((x for x in data if x == target), None)
        if r is not None:
            found += 1
    elapsed = (time.perf_counter() - start) * 1000

    print(f"  find x100 ({size} items): {elapsed}ms, found={found}")

print("BENCHMARK: array_processing")

benchmark_map_filter_reduce(100000)
benchmark_map_filter_reduce(1000000)
benchmark_sort(100000)
benchmark_sort(1000000)
benchmark_find(100000)
benchmark_find(1000000)
